src/Castle.Core/DynamicProxy/CustomAttributeInfo.cs

```diff
@@ -169,14 +170,26 @@ private static Expression UnwrapBody(Expression body)
 
 		private static object GetAttributeArgumentValue(Expression arg, bool allowArray)
 		{
-			var constant = arg as ConstantExpression;
-			if (constant == null)
+			switch (arg.NodeType)
 			{
-				if (allowArray)
-				{
-					var newArrayExpr = arg as NewArrayExpression;
-					if (newArrayExpr != null)
+				case ExpressionType.Constant:
+					return ((ConstantExpression)arg).Value;
+				case ExpressionType.MemberAccess:
+					var memberExpr = (MemberExpression) arg;
+					if (memberExpr.Member is FieldInfo field)
+					{
+						if (memberExpr.Expression is ConstantExpression constant &&
+						    IsCompilerGenerated(constant.Type) &&
```

Comment:
What's the benefit of checking this? I understand that it will catch some cases in which the values aren't of the proper type, but would those cases not be caught when we try to create the attribute anyhow?
Or even be caught by the compiler because we shouldn't have a constructor that accepts the type?
I am especially critical because we're micro-optimizing for the `NodeType` check, as @stakx asked for (and I do like that), but the `IsDefined` call inside `IsCompilerGenerated` is very expensive.

Reply:
> What's the benefit of checking this? I understand that it will catch some cases in which the values aren't of the proper type

No, it checks that the field is a member of a closure type, so that only locals can be accessed like this, not other fields. But maybe it's not necessary. @stakx what do you think?

Reply:
> is a member of a closure type

Oh! Thanks for explaining.